crates/spirv-std/src/arch.rs

```diff
@@ -150,3 +150,26 @@ pub unsafe fn vector_insert_dynamic<T: Scalar, V: Vector<T, N>, const N: usize>(
 pub fn kill() -> ! {
     unsafe { asm!("OpKill", options(noreturn)) }
 }
+
+/// Read from the shader clock.
+///
+/// Requires the `SPV_KHR_shader_clock` extension and the `ShaderClockKHR` capability.
+///
+/// See:
+/// <https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_shader_clock.html>
+#[spirv_std_macros::gpu_only]
+#[doc(alias = "OpReadClockKHR")]
+pub unsafe fn read_clock_khr() -> u64 {
+    let mut result: u64 = 0;
+
+    asm! {
+        "%uint = OpTypeInt 32 0",
+        "%scope = OpConstant %uint {scope}",
+        "%result = OpReadClockKHR typeof*{result} %scope",
+        "OpStore {result} %result",
+        result = in(reg) &mut result,
+        scope = const crate::memory::Scope::Subgroup as _,
```

Comment:
Shouldn't this be a const generic, similar to barriers?

Reply:
Yep, using a const generic here makes sense.

https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_KHR_shader_clock.html#_description says:

> The two valid SPIR-V scopes for `OpReadClockKHR` are `Subgroup` and `Device`.

so we probably want to have a new enum with only those two scopes, to limit invalid usage.

Reply:
We can't use a scope as a const generic yet though:

```rust
error: constant expression depends on a generic parameter
   --> crates\spirv-std\src\arch.rs:189:23
    |
189 |         scope = const SCOPE as u32,
    |                       ^^^^^^^^^^^^
    |
    = note: this may fail depending on what value the parameter takes
                ^^^^^^^^^^^^
 ```

and if we add `#![feature(const_evaluatable_checked)]` to the crate we get:

```rust
error: unconstrained generic constant
   --> crates\spirv-std\src\arch.rs:189:23
    |
189 |         scope = const SCOPE as u32,
    |                       ^^^^^^^^^^^^
    |
    = help: try adding a `where` bound using this expression: `where [(); SCOPE as u32]:`
```

Reply:
Check how barriers do this (barrier.rs), they do this exact thing.

(and for some context, arch.rs is not intended to be ergonomic or safe to use, it's supposed to be the building blocks of a good abstraction)

Reply:
That's fair. Hopefully when const generics can implemented a little further we'll be able to ensure (some) correctness though.